### ISSUE TITLE
Error using DistributedDataParallel with ArcFaceLoss

### ISSUE THREAD

**Issue by rgsousa88** [KevinMusgrave/pytorch-metric-learning#306]
Hi,

I'm trying to train a model using ArcFaceLoss using multiple GPUs but I've got this error and I can't find a away to solve this.

My batch_size is 500 and the embedding size is 512. Here are some snippets of code I'm using to train and compute loss.

```
from apex.parallel import DistributedDataParallel as DDP
from apex.parallel import convert_syncbn_model
from apex import amp

"""Model definition, dataset loading..."""
    
    device = torch.device(f"cuda:{gpu}")

    arc_loss = losses.ArcFaceLoss(num_classes=num_classes,embedding_size=512)
    optimizer = torch.optim.SGD([{'params':model.parameters()},{'params':arc_loss.parameters()}], lr=1e-1, momentum=0.9)
    model.cuda(device)
    arc_loss.cuda(device)
    model = convert_syncbn_model(model)
    arc_loss = convert_syncbn_model(arc_loss)
    [model, arc_loss], optimizer = amp.initialize([model, arc_loss], optimizer, opt_level="O1")
    model = DDP(model)
    arc_loss = DDP(arc_loss)

"""Inside train loop"""
    for epoch in np.arange(epochs):
        model.train()
        for i,data in enumerate(train_loader):            
            images = data[0]['images']
            labels = data[0]['labels'].long()

            optimizer.zero_grad()
            embeddings = model(images)
            loss = arc_loss(embeddings,labels)
```

Error message:

>    Traceback (most recent call last):
>   File "torch_train.py", line 328, in <module>
>     mp.spawn(train, nprocs=args['gpus'], args=(args,))
>   File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 247, in spawn
>     return start_processes(fn, args, nprocs, join, daemon, start_method='spawn')
>   File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 205, in start_processes
>     while not context.join():
>   File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 166, in join
>     raise ProcessRaisedException(msg, error_index, failed_process.pid)
> torch.multiprocessing.spawn.ProcessRaisedException: 
> 
> -- Process 1 terminated with the following error:
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
>     fn(i, *args)
>   File "/home/work/myuserID/code/torch_train.py", line 230, in train
>     loss = arc_loss(embeddings,labels)
>   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
>     result = self.forward(*input, **kwargs)
>   File "/opt/conda/lib/python3.6/site-packages/apex/parallel/distributed.py", line 560, in forward
>     result = self.module(*inputs, **kwargs)
>   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
>     result = self.forward(*input, **kwargs)
>   File "/home/work/.local/lib/python3.6/site-packages/pytorch_metric_learning/losses/base_metric_loss_function.py", line 34, in forward
>     loss_dict = self.compute_loss(embeddings, labels, indices_tuple)
>   File "/home/work/.local/lib/python3.6/site-packages/pytorch_metric_learning/losses/large_margin_softmax_loss.py", line 115, in compute_loss
>     logits = cosine + (mask * diff)
>   File "/opt/conda/lib/python3.6/site-packages/apex/amp/wrap.py", line 58, in wrapper
>     return orig_fn(*new_args, **kwargs)
> RuntimeError: The size of tensor a (500) must match the size of tensor b (1500) at non-singleton dimension 0
> 

Thanks for any help.

Best regards.

**Comment by KevinMusgrave** [comment]
Can you print the shape of ```embeddings``` and ```labels```?

```python
print(embeddings.shape, labels.shape)
loss = arc_loss(embeddings,labels)
```

**Comment by rgsousa88** [comment]
Yes, I already have verified them. The shapes are: embedding -> (500,512)
labels -> (500,1)

**Comment by KevinMusgrave** [comment]
Can you remove that 2nd dimension in labels?

```python
labels = labels.squeeze(1)
```

**Comment by rgsousa88** [comment]
It works! Thanks a lot!
Now, I'm facing other problem related to loss with "nan" value which is crashing the scaled_loss.backward call. Do you think that is related some way 2nd label dimension removal?

**Comment by KevinMusgrave** [comment]
Hmm not sure. A couple of things that might help in debugging:

- set the ArcFaceLoss ```scale``` to 1, like this: ```losses.ArcFaceLoss(num_classes=num_classes,embedding_size=512, scale=1)```
- instead of ArcFaceLoss, try torch.nn.CrossEntropyLoss(). You'll need to add a Linear layer after your existing model.

**Comment by rgsousa88** [comment]
I had no problem training a model using ArcFaceLoss in a single GPU environment. I already have a model trained using CrossEntropyLoss and I was investigating if ArcFaceLoss could perform better. Maybe the problem now is related to my multi gpu implementation... I will keep investigating.

Thanks for your help and advices, @KevinMusgrave.

**Comment by KevinMusgrave** [comment]
If you are using mixed precision, you may want to install the latest pre-release. It contains a bug-fix related to ArcFaceLoss and mixed precision.

```
pip install pytorch-metric-learning==0.9.99.dev0
```

**Comment by rgsousa88** [comment]
Hi, I've  upgraded the module version but I still can't train using 2 GPUs. I get the following error message... I still can't find any solution or help for this error (which might not be related to this module).

> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
>     fn(i, *args)
>   File "/home/work/myuserID/code/torch_train.py", line 248, in train
>     scaled_loss.backward()
>   File "/opt/conda/lib/python3.6/site-packages/torch/tensor.py", line 227, in backward
>     torch.autograd.backward(self, gradient, retain_graph, create_graph)
>   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 138, in backward
>     allow_unreachable=True)  # allow_unreachable flag
>   File "/opt/conda/lib/python3.6/site-packages/apex/parallel/distributed.py", line 400, in allreduce_hook
>     self.comm_ready_buckets(param)
>   File "/opt/conda/lib/python3.6/site-packages/apex/parallel/distributed.py", line 519, in comm_ready_buckets
>     bucket_idx, bucket_loc = self.param_id_to_bucket[id(param)]
> KeyError: 139892655707984

**Comment by KevinMusgrave** [comment]
This might be related: https://github.com/NVIDIA/apex/issues/149

**Comment by KevinMusgrave** [comment]
You mentioned that you already trained a model with CrossEntropyLoss. Was that across multiple GPUs?

**Comment by rgsousa88** [comment]
Hi, Kevin. Sorry for taking so long to reply your question... About training with CrossEntropyLoss, at first I've trained with a single GPU and them a tried with multiple ones and it succeeded.
To solve the "KeyError" problem I took the suggestion to set **_delay_allreduce_** to True in DistributedDataParallel (from apex) constructor. 
But other problem happened (process killed after few iterations) and after I've talked with some colleagues, we figured out that the new problem was more related to environment settings than the model or loss function itself.

Thanks a lot for your help!
